script/hardhat/fork/vip-framework/utils.ts

```diff
@@ -23,15 +23,20 @@ export async function setForkBlock(blockNumber: number) {
 
 export function getCalldatas({ signatures, params }: { signatures: string[]; params: any[][] }) {
   return params.map((args: any[], i: number) => {
-    const types = getArgs(signatures[i]);
+    let types = getArgs(signatures[i]);
+    // Fix for the diamond VIP as there is struct in types and defaultAbiCoder
+    // is unable to process struct.
+    if (signatures[i] == "diamondCut((address,uint8,bytes4[])[],address,bytes)") {
+      types = ["tuple(address, uint8, bytes4[])[]", "address", "bytes"];
+    }
```

Comment:
You can handle this in a more generic way by making a fragment out of the signature using ethers.js, then reading each input to get its type (and handle each of them, including tuples).
You can see that done on the frontend in this PR: https://github.com/VenusProtocol/venus-protocol-interface/pull/1023/files

Reply:
I guess we cannot implement it here and there is only one use case here.

Reply:
The solution I linked could be implemented here since it uses plain JavaScript.
@chechu told me that another case was found by Narayan who used a similar approach to yours, hence why he suggested using a more generic solution

Reply:
You can get the types of the params from the signature. If you check the PR I linked, you'll see this process being done.
You first pass the signature to `ethers.utils.FunctionFragment.from`:
```typescript
const fragment = ethers.utils.FunctionFragment.from(
  signatures[i], // diamondCut((address,uint8,bytes4[])[],address,bytes)
);
```

The `fragment` variable will look like this:
<img width="532" alt="Screen Shot 2023-04-20 at 17 09 51" src="https://user-images.githubusercontent.com/44675210/233425436-6291e24b-63d8-4682-bf5d-69bbe880b216.png">

As you can see, you get an object with an `inputs` property that describes the type of each param. You can iterate other this property, which is an array, to then form your final string representing the types. Here's the code that would do that:

```typescript
import { ParamType } from 'ethers/lib/utils';

const formatParamType = (paramType: ParamType): string => {
  if (paramType.type !== 'tuple' && paramType.type !== 'tuple[]') {
    return paramType.type;
  }

  let res = `tuple(${paramType.components.map(formatParamType).join(', ')})`;

  if (paramType.type === 'tuple[]') {
    res += '[]';
  }

  return res;
};

const types = fragment.inputs.map(formatParamType);
```

This logic would replace the `getArgs` function you are currently using and should be generic enough to handle all types of signatures.

Reply:
Thanks @therealemjy for the explanation, I have made the changes. Please review it at commit: a1e2f09aca55f45cf2c56c4f45db051b6ad4b010

Reply:
No worries! Looks good to me